apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,270 @@
+		// immutability validation
+		if policy.Immutable && !featureRef.SkipStabilityPolicyValidation && policy.DefaultActivation != featureRef.Activate {
+			invalidFeatures.Insert(featureRef.Name)
+		} else if policy.VoidsWarranty && !featureRef.SkipStabilityPolicyValidation { // warranty validation
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+
+	if len(invalidFeatures) != 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("cannot toggle features due to policy violation: %v."+
+				"Set feature.skipStabilityPolicyValidation to true to override policy."+
+				"Once set to true, cannot be set back to false.", invalidFeatures.List())))
+	}
+	return allErrors
+}
+
+// validateStabilityPolicyOverride validates stability policy override
+func (r *FeatureGate) validateStabilityPolicyOverride(oldObject *FeatureGate) field.ErrorList {
+	var allErrors field.ErrorList
+
+	invalidFeatures := sets.String{}
+	for _, featureRef := range r.Spec.Features {
+		skipStabilityPolicyValidationFlagFromOldObj, found := getSkipStabilityPolicyValidationFieldForFeature(oldObject, featureRef.Name)
+		if !found {
+			// Feature doesn't exist in old object and need not be validated
+			continue
+		}
+		if !featureRef.SkipStabilityPolicyValidation && skipStabilityPolicyValidationFlagFromOldObj {
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+	if len(invalidFeatures) != 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("cannot set the feature.skipStabilityPolicyValidation back to false after the stability "+
+				"policy has been overridden: %v", invalidFeatures.List())))
+	}
+	return allErrors
+}
+
+// validateFeatureExistence validates feature existence in the cluster
+func (r *FeatureGate) validateFeatureExistence(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	allFeaturesInSpec := sets.String{}
+	for _, feature := range r.Spec.Features {
+		allFeaturesInSpec.Insert(feature.Name)
+	}
+
+	allFeaturesInCluster := sets.String{}
+	for _, feature := range features.Items {
+		allFeaturesInCluster.Insert(feature.Name)
+	}
+
+	invalidFeatures := allFeaturesInSpec.Difference(allFeaturesInCluster)
+
+	if len(invalidFeatures) != 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("some features in the FeatureGate spec do not exist in cluster: %v", invalidFeatures.List())))
+	}
+	return allErrors
+}
+
+// validateFeatureConflicts validates that the feature is only being gated by one featuregate
+func (r *FeatureGate) validateFeatureConflicts(ctx context.Context, c client.Client) field.ErrorList {
```

Comment:
Thanks to the helpful comment, I think the function's name would be better described as `validateFeatureGateConflict`. Because if you have more than one FeatureGate for one feature, it's not the feature that conflicts, is it? It's the FeatureGate that will conflict.

Reply:
i updated the method to `validateConflictingFeaturesInFeatureGate` and i think  it makes more sense